libmamba/include/mamba/core/problems_graph_util.hpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_PROBLEMS_GRAPH_HPP
+#define MAMBA_PROBLEMS_GRAPH_HPP
+
+#include <string>
+#include <utility>
+#include <optional>
+#include <vector>
+#include <regex>
+#include <variant>
+#include <unordered_set>
+#include <sstream>
+
+#include "mamba/core/graph_util.hpp"
+#include "mamba/core/package_info.hpp"
+
+extern "C"
+{
+#include "solv/solver.h"
+}
+
+namespace mamba
+{
+
+    enum class ProblemType
+    {
+        CONFLICT,
+        NOT_FOUND,
+        NOT_INSTALLABLE,
+        BEST_NOT_INSTALLABLE,
+        ONLY_DIRECT_INSTALL,
+        EXCLUDED_BY_REPO_PRIORITY,
+        INFERIOR_ARCH,
+        PROVIDED_BY_SYSTEM
+    };
+
+    class DependencyInfo
+    {
+    public:
+        DependencyInfo(const std::string& dep);
+
+        const std::string& get_name() const;
+        const std::string& get_range() const;
+        std::string str() const;
+
+    private:
+        std::string name;
+        std::string range;
+    };
+
+    namespace NodeInfo
+    {
+        struct ResolvedPackage
+        {
+            PackageInfo m_package_info;
+            ResolvedPackage(PackageInfo const& package_info);
+        };
+
+        struct ProblematicPackage
+        {
+            std::string m_dep;
```

Comment:
Yes, it would be the same dependency.  I wanted to be consistent in terms of what we have in the node so we can explain the errors when parsing the graph. But it should be ok to have it as an empty class